src/vec_int128_ppc.h

```diff
@@ -608,6 +610,287 @@ test_mul4uq (vui128_t *__restrict__ mulu, vui128_t m1h, vui128_t m1l,
+ * This is a bit tricky as the quotient is normally passed by value,
+ * but for this operation we need to pass by reference,
+ * which allows the corrected quotient to be passed on to the next step.
+ *
+ * Again as this operation will be used in a loop for long division
+ * operations and the extended multiplies are fairly expensive,
+ * we should check for and short-circuit special conditions.
+ * The most important special condition is when the dividend is less
+ * that the divisor and the remainder is simply the dividend.
+ *
+ * \code
+static inline vui128_t
+vec_modudq_10e31 (vui128_t vra, vui128_t vrb, vui128_t *ql)
+{
+  // ten31  = +100000000000000000000000000000000UQ
+  const vui128_t ten31 = (vui128_t)
+	  { (__int128) 1000000000000000UL * (__int128) 10000000000000000UL };
+  const vui128_t zero = (vui128_t) { (__int128) 0UL };
+  const vui128_t minus_one = (vui128_t) { (__int128) -1L };
+  vui128_t result, t, th, c;
+
+  if (vec_cmpuq_all_ne (vra, zero) || vec_cmpuq_all_ge (vrb, ten31))
+    {
+      t = vec_muludq (&th, *ql, ten31);
+      c = vec_subcuq (vrb, t);
+      t = vec_subuqm (vrb, t);
+      th = vec_subeuqm (vra, th, c);
+      // The remainder should be less than the divisor
+      if (vec_cmpuq_all_ne (th, zero) && vec_cmpuq_all_ge (t, ten31))
+	{
+	  // If not the estimated quotient is off by 1
+	  *ql = vec_adduqm (*ql, minus_one);
+	  // And the remainder is negative, so add the divisor
+	  t = vec_adduqm (t, ten31);
+	}
+      result = t;
+    }
+  else
+    result = vrb;
+
+  return result;
+}
+ * \endcode
+ *
+ * Now we have all the operations needed to complete the implementation
+ * of long division by the decimal constant (10<SUP>31</SUP>).
+ *
+ * \code
+vui128_t
+example_longdiv_10e31 (vui128_t *q, vui128_t *d, long int _N)
+{
+  vui128_t dn, qh, ql, rh;
+  long int i;
+
+  // initial step for the top digits
+  dn = d[0];
+  qh = vec_divuq_10e31 (dn);
+  rh = vec_moduq_10e31 (dn, qh);
+  q[0] = qh;
+
+  // now we know the remainder is less than the divisor.
+  for (i=1; i<_N; i++)
+    {
+      dn = d[i];
+      ql = vec_divudq_10e31 (&qh, rh, dn);
+      rh = vec_modudq_10e31 (rh, dn, &ql);
+      q[i] = ql;
+    }
+  // return the final remainder
+  return rh;
+}
+ * \endcode
+ * The result of each call to example_longdiv_10e31() is the output
+ * array <I>q</I> of quadwords containing the extended quotient,
+ * and the remainder as the return value.
+ * The input array <I>d</I> and output array <I>q</I> should not
+ * overlap in storage.
+ * The remainder is in the range 0-9999999999999999999999999999999
+ * and is suitable for conversion to BCD or decimal characters.
+ * (see vec_bcdcfsq()).
+ * Repeated calls passing the quotient from the previous call as the dividend,
+ * reduces the quotient by 31 digits returns another 31 digits in the remainder for conversion.
+ * This continues until the quotient is less than 10<SUP>31</SUP>
+ * and the provides the high order 31-digits of the decimal result.
```

Comment:
s/the/thus/ ?  or then?

Reply:
lets try:

 * This continues until the quotient is less than 10<SUP>31</SUP>
 * which provides the highest order digits of the decimal result.